### PR TITLE
[rllib] Ape-X implementation and DQN refactor to handle replay in policy optimizer

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This implements Ape-X Distributed Prioritized Experience Replay (https://openreview.net/forum?id=H1Dy---0Z ) as a separate policy optimizer module for DQN.

There were a couple changes in DQN as well:
- move replay buffer handling into the policy optimizer
- remove the multi-gpu and async support (it never performed that well, and apex should supercede these modes)
- add snappy compression support. This reduces the storage per observation from ~100KB to ~1.5KB. Snappy can decompress at ~500MB/s, which is fast enough to not be noticeable in normal DQN. In Ape-X, we decompress asynchronously in the replay buffer shards so it isn't in the critical path of the learner thread.

(perf results will be included in a separate pr which swaps the current preprocessor out in favor of the deepmind wrapper)

## Related issue number

https://github.com/ray-project/ray/issues/1395

### CODE REVIEW

File: python/ray/rllib/dqn/dqn_evaluator.py [hunk 1]
```diff
@@ -112,33 +119,48 @@ def sample(self):
...
+
         return batch
 
     def compute_gradients(self, samples):
-        _, grad = self.dqn_graph.compute_gradients(
+        td_error, grad = self.dqn_graph.compute_gradients(
             self.sess, samples["obs"], samples["actions"], samples["rewards"],
             samples["new_obs"], samples["dones"], samples["weights"])
-        return grad
+        return grad, td_error
```
Comment (richardliaw) [inline python/ray/rllib/dqn/dqn_evaluator.py]: I don't think we need to actually return this here right?

File: python/ray/rllib/utils/window_stat.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+
+class WindowStats(object):
```
Comment (richardliaw) [inline python/ray/rllib/utils/window_stat.py]: preferably `WindowStat` for consistency

File: python/ray/rllib/optimizers/apex_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+            ev.set_weights.remote(weights)
+            self.steps_since_update[ev] = 0
+            for _ in range(SAMPLE_QUEUE_DEPTH):
+                self.sample_tasks.add(ev, ev.sample.remote())
+
+    def step(self):
+        start = time.time()
+        sample_timesteps, train_timesteps = self._step()
+        time_delta = time.time() - start
+        self.sample_timer.push(time_delta)
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/optimizers/apex_optimizer.py]: Hm there's this weird overloading of `TimerStat` to measure throughput. 

I wonder if there's a more elegant way to do this - perhaps have a separate class that extends TimerStat? and then within the with statement, you can record via `stat.push_units(sample_timesteps)`. So something like:

```
sample_stat, train_stat = StreamStat(), StreamStat()
with sample_stat, train_stat:
    sample_timesteps, train_timesteps = self._step()
    sample_stat.push_units(sample_timesteps)
    train_stat.push_units(train_timesteps)
if train_timesteps == 0 and not self.learning_started:
    train_stat.clear()
else:
    self.learning_started = True
return sample_timesteps
```

File: python/ray/rllib/utils/timer.py [hunk 4]
```diff
@@ -4,10 +4,10 @@
 
 import time
 
-from ray.rllib.utils.filter import RunningStat
+import numpy as np
 
 
-class TimerStat(RunningStat):
+class TimerStat(object):
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/utils/timer.py]: What about keeping TimerStat and extending TimerStat functionality in a subclass `StreamStat`

```
class StreamStat(TimerStat, WindowStat):
    
    def __init__(self, window_size=10)
    def __enter__()
    def __exit__()
    def push(time)
    def push_units(units)
    
    @property
    def mean_time()

    @property
    def mean_units()

    @property
    def total_time()

    @property
    def total_units()
    
    @property
     def mean_throughput()

File: python/ray/rllib/optimizers/apex_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+import ray
+from ray.rllib.dqn.replay_buffer import ReplayBuffer, PrioritizedReplayBuffer
+from ray.rllib.optimizers.optimizer import Optimizer
+from ray.rllib.optimizers.sample_batch import SampleBatch
+from ray.rllib.utils.actors import TaskPool, create_colocated
+from ray.rllib.utils.timer import TimerStat
+from ray.rllib.utils.window_stat import WindowStats
+
+SAMPLE_QUEUE_DEPTH = 2
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/apex_optimizer.py]: why not expose these as optimizer configs?

**Review by richardliaw — COMMENTED** [review]
main comment is the timer code is pretty unappealing; posted a suggestion for you to consider. Actor utilities are nice, and the overall DQN cleanup is really neat.

It may also make some sense to consider splitting Ape-X from DQN as a separate algorithm.

File: python/ray/rllib/dqn/dqn_evaluator.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/dqn/dqn_evaluator.py]: You do if you use another optimizer

File: python/ray/rllib/utils/timer.py [hunk 6]
```diff
@@ -4,10 +4,10 @@
...
     """A running stat for conveniently logging the duration of a code block.
 
     Example:
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/timer.py]: Fixed

File: python/ray/rllib/utils/timer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/timer.py]: Hmm maybe we can do this in another PR if needed.

**Review by ericl — COMMENTED** [review]
Split out into APEX agent, did refactorings of timers.

File: python/ray/rllib/dqn/dqn_evaluator.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/rllib/dqn/dqn_evaluator.py]: which optimizer requires this? i thought all other optimizers only return grad

Comment (ericl) [reply python/ray/rllib/dqn/dqn_evaluator.py]: Ah I guess theoretically a variant of local_sync_replay that doesn't use the fused compute_apply(). For now, I guess we can remove these two methods entirely.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**